ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InMemoryKVParticipantState.scala

```diff
@@ -0,0 +1,216 @@
+      // Submitter chosen entry identifier
+      entryId: DamlLogEntryId,
+      submission: DamlSubmission
+  )
+}
+
+/* Implementation of the participant-state using the key-value utilities and an in-memory map. */
+class InMemoryKVParticipantState(implicit system: ActorSystem)
+    extends ReadService
+    with WriteService
+    with AutoCloseable {
+  import InMemoryKVParticipantState._
+
+  val ledgerId = SimpleString.assertFromString(UUID.randomUUID.toString)
+
+  // DAML Engine for transaction validation.
+  private val engine = Engine()
+
+  // Random number generator for generating unique entry identifiers.
+  private val rng = new java.util.Random
+
+  // Namespace prefix for log entries.
+  private val NS_LOG_ENTRIES = ByteString.copyFromUtf8("L")
+
+  // Namespace prefix for DAML state.
+  private val NS_DAML_STATE = ByteString.copyFromUtf8("DS")
+
+  // Reference to the latest immutable state. Reference is only updated by the CommitActor.
+  // Reading from the state must happen by first taking the reference (val state = stateRef).
+  @volatile private var stateRef: State =
+    State(
+      log = Vector.empty[(DamlLogEntryId, DamlSubmission)],
+      recordTime = Timestamp.Epoch,
+      store = Map.empty[ByteString, ByteString],
+      config = Configuration(
+        timeModel = TimeModel.reasonableDefault
+      )
+    )
+
+  // Akka actor that receives submissions and commits them to the key-value store.
+  @SuppressWarnings(Array("org.wartremover.warts.Any"))
+  class CommitActor extends Actor {
+    override def receive: Receive = {
+      case CommitMessage(entryId, submission) =>
+        val state = stateRef
+        if (state.store.contains(entryId.getEntryId)) {
+          // The entry identifier already in use, drop the message and let the
+          // client retry submission.
+        } else {
+          // Process the submission to produce the log entry and the state updates.
+          val (logEntry, damlStateUpdates) = KeyValueCommitting.processSubmission(
+            engine,
+            state.config,
+            entryId,
+            state.recordTime,
+            submission,
+            submission.getInputLogEntriesList.asScala
+              .map(eid => eid -> getLogEntry(state, eid))(breakOut),
+            submission.getInputDamlStateList.asScala
+              .map(key => key -> getDamlState(state, key))(breakOut)
+          )
+
+          // Combine the abstract log entry and the state updates into concrete updates to the store.
+          val allUpdates =
+            damlStateUpdates.map {
+              case (k, v) =>
+                NS_DAML_STATE.concat(KeyValueCommitting.packDamlStateKey(k)) ->
+                  KeyValueCommitting.packDamlStateValue(v)
+            } + (entryId.getEntryId -> KeyValueCommitting.packDamlLogEntry(logEntry))
+
+          // Atomically update the state reference, recording the commit
+          // and the updated state.
+          stateRef = state.copy(
```

Comment:
The way this is written it seems to me that we MUST re-validate the read input values on the update here! Otherwise, we are racy!

Reply:
Or explain why we are not in case the actor guards us. I don't understand the 'atomically' in that case though? Is it wrt readers only seeing properly updated state?